core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java

```diff
@@ -251,13 +253,25 @@ public static RelDataType toSql(final RelDataTypeFactory typeFactory,
               type.getFieldNames()),
           type.isNullable());
     } else if (type instanceof JavaType) {
-      return typeFactory.createTypeWithNullability(
-          typeFactory.createSqlType(type.getSqlTypeName()),
-          type.isNullable());
+      Class javaClass = ((JavaType) type).getJavaClass();
+      final RelDataType nonNullType;
+      if (javaClass.isArray()) {
+        nonNullType = typeFactory.createArrayType(type.getComponentType(), -1);
+      } else if (isCollection(javaClass)) {
+        nonNullType = typeFactory.createArrayType(typeFactory.createSqlType(SqlTypeName.ANY), -1);
+      } else {
+        nonNullType = typeFactory.createSqlType(type.getSqlTypeName());
+      }
+      return typeFactory.createTypeWithNullability(nonNullType, type.isNullable());
     }
     return type;
   }
 
+  private static boolean isCollection(Class javaClass) {
+    return javaClass == Collection.class
+        || Arrays.stream(javaClass.getInterfaces()).anyMatch(JavaTypeFactoryImpl::isCollection);
+  }
```

Comment:
This would form a n * n *.. recursion, i don't think we need that, for each super class, only `javaClass == Collection.class` is enough.